raiden-dapp/tests/unit/components/mixins/ethereum-provider-dialog-mixin.spec.ts

```diff
@@ -0,0 +1,164 @@
+import { DirectRpcProvider } from '@/services/ethereum-provider';
+
+jest.mock('@/services/ethereum-provider/direct-rpc-provider');
+
+Vue.use(Vuex);
+
+@Component({ template: '<div></div>' })
+class TestComponent extends Mixins(EthereumProviderDialogMixin) {
+  providerFactory = DirectRpcProvider;
+  rpcUrl = '';
+  privateKey = '';
+}
+
+const saveEthereumProviderOptions = jest.fn();
+const getEthereumProviderOptions = jest.fn().mockReturnValue({});
+const mutations = { saveEthereumProviderOptions };
+const getters = { getEthereumProviderOptions: () => getEthereumProviderOptions };
+const userSettings = {
+  namespaced: true,
+  getters,
+  mutations,
+};
+
+function createWrapper(options?: {
+  linkingInProgress?: boolean;
+  canLink?: boolean;
+  providerOptions?: unknown;
+}): Wrapper<TestComponent> {
+  const store = new Vuex.Store({
+    modules: { userSettings },
+  });
+
+  const wrapper = shallowMount(TestComponent, { store });
+  wrapper.setData({ linkingInProgress: options?.linkingInProgress ?? false });
+
+  if (options?.canLink !== undefined) {
+    Object.defineProperty(wrapper.vm, 'canLink', { get: () => options.canLink });
+  }
+
+  if (options?.providerOptions !== undefined) {
+    Object.defineProperty(wrapper.vm, 'providerOptions', { get: () => options.providerOptions });
+  }
+
+  return wrapper;
+}
+
+describe('EthereumProviderDialogMixin', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  test('enables linking per default', () => {
+    const wrapper = createWrapper();
+
+    expect(wrapper.vm.canLink).toBeTruthy();
+  });
+
+  test('has empty provider options per default', () => {
+    const wrapper = createWrapper();
+
+    expect(wrapper.vm.providerOptions).toMatchObject({});
+  });
+
+  test('loads saved provider options from store on creation', () => {
+    getEthereumProviderOptions.mockReturnValueOnce({ rpcUrl: 'testUrl', privateKey: 'testKey' });
+    const wrapper = createWrapper();
+
+    expect(getEthereumProviderOptions).toHaveBeenCalledTimes(1);
+    expect(wrapper.vm.rpcUrl).toBe('testUrl');
+    expect(wrapper.vm.privateKey).toBe('testKey');
+  });
+
+  test('linking fails if not enabled', () => {
+    const wrapper = createWrapper({ canLink: false });
+
+    expect(wrapper.vm.link()).rejects.toThrowError('Can not link now!');
+  });
+
+  test('linking fails if already in progress', () => {
+    const wrapper = createWrapper({ linkingInProgress: true });
+
+    expect(wrapper.vm.link()).rejects.toThrowError('Can not link now!');
+  });
+
+  test('changes internal state to failed if provider factory throws', async () => {
+    (DirectRpcProvider as any).link.mockRejectedValueOnce(new Error('failed'));
+    const wrapper = createWrapper();
+    expect(wrapper.vm.linkingFailed).toBeFalsy();
+
+    await wrapper.vm.link();
+
+    expect(wrapper.vm.linkingFailed).toBeTruthy();
+  });
+
+  test('after linking finishes, in any case it resets the in progress state', async () => {
```

Comment:
`resets the in progress state after linking finishes`

Reply:
✔️